File: Data/Sally/sallySource.py
import numpy as np

###############################################################################
'''Functions'''
###############################################################################

def gen_Sally(xs,ys,zs,time=1.0):

    r2 = 8
    SMALL = 0.00000000001
    xdelta = 1.0 / (xs-1.0)
    ydelta = 1.0 / (ys-1.0)
    zdelta = 1.0 / (zs-1.0)
    
    ix_values = []
    iy_values = []
    iz_values = []

    x_vec = []
    y_vec = []
    z_vec = []
    
    data = []

    for iz in range(0,int(zs)):
        # map z to 0->1
        z = iz * zdelta
        # For each z-slice, determine the spiral circle
        xc = 0.5 + 0.1*np.sin(0.04*time+10.0*z)
        #(xc,yc) determine the center of the circle.
        yc = 0.5 + 0.1*np.cos(0.03*time+3.0*z)
        #The radius also changes at each z-slice.
        r = 0.1 + 0.4 * z*z + 0.1 * z * np.sin(8.0*z)
        r2 = 0.2 + 0.1*z

        for iy in range(0,int(ys)):
            y = iy * ydelta
            
            for ix in range(0,int(xs)):
                x = ix * xdelta
                temp = np.sqrt( (y-yc)*(y-yc) + (x-xc)*(x-xc) )
                scale = np.fabs( r - temp )
                if ( scale > r2 ):
                    scale = 0.8 - scale
                else:
                    scale = 1.0
                z0 = 0.1 * (0.1 - temp*z )
                if ( z0 < 0.0 ):
                    z0 = 0.0
                temp = np.sqrt( temp*temp + z0*z0 )
                scale = (r + r2 - temp) * scale / (temp + SMALL)
                scale = scale / (1+z)
                xi = scale * (y-yc) + 0.1*(x-xc)
                yi = scale * -(x-xc) + 0.1*(y-yc)
                zi = scale * z0
                x_vec.append(xi)
                y_vec.append(yi)
                z_vec.append(zi)
                ix_values.append(ix)
                iy_values.append(iy)
                iz_values.append(iz)
                
    data.append(x_vec)
    data.append(y_vec)
    data.append(z_vec)
    data.append(ix_values)
    data.append(iy_values)
    data.append(iz_values)
    data = np.asarray(data)

    return data.T

###############################################################################
    
def animate_Sally(xs,ys,zs,time,steps):
    
    c = 0
    
    for t in np.linspace(0,time,steps):
        data = gen_Sally(xs,ys,zs, t)
        np.savetxt('Time/data.csv.'+str(c),data, delimiter = ',',
           header = 'x_vec,y_vec,z_vec,ix,iy,iz',
           comments = '')
        c+=1
        print('%.2f' %((c/steps)*100) + '%')
        
###############################################################################
'''Implementation'''
###############################################################################

xs,ys,zs = 20,20,20

time = 1000
steps = 1000

animate_Sally(xs,ys,zs,time,steps)

File: Data/Sally/sallySource.c

#include <math.h>
void gen_sally( int xs, int ys, int zs, int time, float *sally )
/*
 *  Gen_Sally creates a vector field of dimension [xs,ys,zs,3] from
 *  a proceedural function. By passing in different time arguements,
 *  a slightly different and rotating field is created.
 *
 *  The magnitude of the vector field is highest at some funnel shape
 *  and values range from 0.0 to around 0.4 (I think).
 *
 *  I just wrote these comments, 8 years after I wrote the function.
 *  
 *  Developed by Sally of Sally University
 *
 */
{
  float x, y, z;
  int ix, iy, iz;
  float r, xc, yc, scale, temp, z0;
  float r2 = 8;
  float SMALL = 0.00000000001;
  float xdelta = 1.0 / (xs-1.0);
  float ydelta = 1.0 / (ys-1.0);
  float zdelta = 1.0 / (zs-1.0);

  for( iz = 0; iz < zs; iz++ )
  {
     z = iz * zdelta;                        // map z to 0->1
     xc = 0.5 + 0.1*sin(0.04*time+10.0*z);   // For each z-slice, determine the spiral circle.
     yc = 0.5 + 0.1*cos(0.03*time+3.0*z);    //    (xc,yc) determine the center of the circle.
     r = 0.1 + 0.4 * z*z + 0.1 * z * sin(8.0*z); //  The radius also changes at each z-slice.
     r2 = 0.2 + 0.1*z;                           //    r is the center radius, r2 is for damping
     for( iy = 0; iy < ys; iy++ )
     {
		y = iy * ydelta;
		for( ix = 0; ix < xs; ix++ )
		{
			x = ix * xdelta;
			temp = sqrt( (y-yc)*(y-yc) + (x-xc)*(x-xc) );
			scale = fabs( r - temp );
/*
 *  I do not like this next line. It produces a discontinuity 
 *  in the magnitude. Fix it later.
 *
 */
           if ( scale > r2 )
              scale = 0.8 - scale;
           else
              scale = 1.0;
			z0 = 0.1 * (0.1 - temp*z );
		   if ( z0 < 0.0 )  z0 = 0.0;
		   temp = sqrt( temp*temp + z0*z0 );
			scale = (r + r2 - temp) * scale / (temp + SMALL);
			scale = scale / (1+z);
           *sally++ = scale * (y-yc) + 0.1*(x-xc);
           *sally++ = scale * -(x-xc) + 0.1*(y-yc);
           *sally++ = scale * z0;
		}
     }
  }
}
